### ISSUE TITLE
CUDA: step time grows with each new tensor shape — cached metadata buffers make every allocation O(live slices)

### ISSUE THREAD

**Issue by swfsql** [tracel-ai/burn#5751]
- *Analysis and reproduction prepared by Claude.*
- *The base-case was the training getting slower over time (up to a ceiling, but still permanent and real).*

---

**Describe the bug**

In a training loop whose tensor shapes change from step to step — a batch of
variable-length sequences is the ordinary way to get there — the **host** time of every
step grows with the number of steps taken, and never comes back down. A fresh process is
fast again. VRAM and host RSS stay flat the whole time, so it does not look like a leak.

Two mechanisms multiply:

1. **The metadata info cache fills with single-use entries.** Every launch's info buffer is
   cached keyed by the exact info words — shapes, strides and scalars
   ([`InfoCacheKey`](https://github.com/tracel-ai/cubecl/blob/2a3c997dafc43081d58c1417111c35c726dc46d6/crates/cubecl-server/src/metadata_cache.rs#L55)) —
   LRU-capped at **4096 entries** of ≤2048 B
   ([`MetadataCachePolicy::default`](https://github.com/tracel-ai/cubecl/blob/2a3c997dafc43081d58c1417111c35c726dc46d6/crates/cubecl-server/src/metadata_cache.rs#L85);
   the CUDA and HIP streams both take that default, wgpu uses `new(512, 2048)`).
   A workload whose shapes keep changing misses on nearly every launch, so the cache fills
   with entries that are never read again, and each entry holds one **live slice** in a
   dynamic memory pool for as long as it stays cached — 512 B on CUDA, where the device
   alignment rounds every small reservation up to that.

2. **A reservation costs O(live slices in the pool that serves its size).**
   [`SlicedPool::try_reserve`](https://github.com/tracel-ai/cubecl/blob/2a3c997dafc43081d58c1417111c35c726dc46d6/crates/cubecl-server/src/memory_management/memory_pool/sliced_pool.rs#L153)
   walks the pool's pages in order until one fits, and on each page it visits:
   [`MemoryPage::coalesce`](https://github.com/tracel-ai/cubecl/blob/2a3c997dafc43081d58c1417111c35c726dc46d6/crates/cubecl-server/src/memory_management/memory_pool/memory_page.rs#L251)
   scans every slice for runs of adjacent free ones (an `is_free` per slice) and, when it
   finds one, rebuilds the page's slice vector, calling `update_slice` on **every**
   surviving slice; then
   [`MemoryPage::try_reserve`](https://github.com/tracel-ai/cubecl/blob/2a3c997dafc43081d58c1417111c35c726dc46d6/crates/cubecl-server/src/memory_management/memory_pool/memory_page.rs#L166)
   scans the slices for a first fit, and
   [`add_new_slice`](https://github.com/tracel-ai/cubecl/blob/2a3c997dafc43081d58c1417111c35c726dc46d6/crates/cubecl-server/src/memory_management/memory_pool/memory_page.rs#L308)
   rebuilds the vector once more, renumbering **every** slice on the page, not just the ones
   after the split. This runs on every
   [`MemoryManagement::reserve`](https://github.com/tracel-ai/cubecl/blob/2a3c997dafc43081d58c1417111c35c726dc46d6/crates/cubecl-server/src/memory_management/memory_manage.rs#L636).

The two meet in one pool. A reservation takes the first pool that accepts its size, and on
this card the `SubSlices` preset lays out (cubecl's own `[memory.logger]` output):

```
 - Exclusive Pool max_alloc_size=0 B
 - Sliced Pool page_size=24.24 MB max_alloc_size=3.03 MB
 - Sliced Pool page_size=96.97 MB max_alloc_size=24.24 MB
 - Sliced Pool page_size=387.90 MB max_alloc_size=193.95 MB
 - Exclusive Pool max_alloc_size=1.55 GB
```

So the 512 B info buffers *and* every activation up to 3.03 MB are served by the same
sliced pool: the cache steadily adds live slices there, and each added slice makes every
later allocation in that pool slower. The cost lands on the thread that
allocates, which is the device thread, and the calling thread then blocks enqueueing to it
— so forward, backward and the optimizer step all slow down together.

The growth stops only when shapes start repeating (cache hits) or when the cache reaches
4096 entries. In a real training run neither happens quickly: at ~3 new entries per step it
took ~1200 steps (~7 min) to reach the cap in my case, which reads as "it just keeps
getting slower".

**Where this came from, and what it is for.** The info cache is recent: it arrived with
[#1415](https://github.com/tracel-ai/cubecl/pull/1415) (2026-07-09, "Feat/hip captured
graph"), and the feature it serves is one I actively want. A device allocation inside a
capture window is illegal, so every info buffer a captured launch touches has to be warm
before the window opens; caching them is what makes a stable-shape capture clean, and in
normal operation it also saves an allocation and a host→device copy per launch. For small
models — where per-launch host overhead is the dominant cost, which is exactly the regime
this report is about — graph capture is the win, not the problem. So this is a note about an
unintended consequence on the *non-capture* path, not a request to revert anything.

And the capture path does not depend on the policy that causes this. Capture gets what it
needs from [`CacheMode::Capture`](https://github.com/tracel-ai/cubecl/blob/2a3c997dafc43081d58c1417111c35c726dc46d6/crates/cubecl-server/src/metadata_cache.rs#L60),
which covers the whole prepare → record window (warmup included — that is what
`the_cache_captures_across_the_whole_window` pins down): there the cache is unbounded,
evicts nothing, and pins every entry a recording touches to the graph. That holds even at
`max_entries = 0`, since `should_cache` returns `true` in capture mode regardless and
`insert` skips the capacity check entirely when `capacity()` is `None` — so warmup still
primes every buffer the recorded launches will hit. The 4096-entry retention is a separate,
unconditional `CacheMode::Normal` policy, and it is the one a shape-varying workload pays
for with no capture anywhere in sight. Shrinking it, or making it adaptive, costs graph
capture nothing.

**To Reproduce**

A standalone crate that depends only on burn (`3a93fbfc`, cubecl `2a3c997`). The model,
the data and the work per step are identical between runs — the only difference is that
`VARY=1` scores a number of rows that differs from step to step, as a variable-length
batch would.

`Cargo.toml`:

```toml
[package]
name = "slowdown"
version = "0.1.0"
edition = "2024"

[workspace]

[dependencies.burn]
git = "https://github.com/tracel-ai/burn.git"
rev = "3a93fbfc0f6517c9165031f8b2fbbe8523ff5c29"
default-features = false
features = ["autodiff", "cuda", "std", "optim"]
```

`src/main.rs`:

```rust
use burn::nn::loss::{CrossEntropyLoss, CrossEntropyLossConfig};
use burn::nn::{Linear, LinearConfig, Relu};
use burn::optim::{AdamWConfig, GradientsParams};
use burn::prelude::*;
use burn::tensor::Distribution;
use std::time::Instant;

fn env(name: &str, default: usize) -> usize {
    std::env::var(name).map_or(default, |v| v.parse().expect(name))
}

/// Live device allocations `(count, bytes)`, as cubecl sees them.
fn device_memory(device: &Device) -> Option<(u64, u64)> {
    match device.as_dispatch().clone().inner() {
        burn::backend::DispatchDevice::Cube(cube) => {
            let usage = cube.client().memory_usage();
            Some((usage.number_allocs, usage.bytes_in_use))
        }
        #[allow(unreachable_patterns)]
        _ => None,
    }
}

#[derive(Module, Debug)]
struct Mlp {
    layers: Vec<Linear>,
    activation: Relu,
}

impl Mlp {
    fn new(width: usize, layers: usize, device: &Device) -> Self {
        Self {
            layers: (0..layers).map(|_| LinearConfig::new(width, width).init(device)).collect(),
            activation: Relu::new(),
        }
    }

    fn forward(&self, x: Tensor<2>) -> Tensor<2> {
        self.layers.iter().fold(x, |x, layer| self.activation.forward(layer.forward(x)))
    }
}

fn main() {
    let (vary, cycle) = (env("VARY", 1) == 1, env("CYCLE", 512));
    let (steps, every) = (env("STEPS", 1200), env("EVERY", 100));
    let (rows, width, layers) = (env("BATCH", 2048), env("WIDTH", 256), env("LAYERS", 8));

    let device = Device::default().autodiff();
    println!("device {device:?}; vary {vary} cycle {cycle}");
    device.seed(0);
    let mut model = Mlp::new(width, layers, &device);
    let mut optim = AdamWConfig::new().init();

    let inputs = Tensor::<2>::random([rows, width], Distribution::Default, &device);
    let targets = Tensor::<2>::random([rows, width], Distribution::Default, &device)
        .argmax(1)
        .squeeze_dim::<1>(1);
    let loss_fn: CrossEntropyLoss = CrossEntropyLossConfig::new().init(&device);

    let start = Instant::now();
    let mut sum = 0.0;
    for step in 1..=steps {
        let t0 = Instant::now();
        let output = model.forward(inputs.clone());
        // The one varying shape: how many rows the loss scores.
        let scored = match vary {
            true => rows - step % cycle.min(rows),
            false => rows,
        };
        let loss = loss_fn.forward(
            output.narrow(0, 0, scored),
            targets.clone().narrow(0, 0, scored),
        );
        let grads = GradientsParams::from_grads(loss.backward(), &model);
        model = optim.step(1e-4, model, grads);
        sum += (Instant::now() - t0).as_secs_f64() * 1e3;

        if step % every == 0 {
            let (allocs, bytes) = device_memory(&device).unwrap_or_default();
            println!(
                "step {step:>5} t={:>3.0}s  {:>7.2} ms/step  allocs {allocs:>5} ({:.1} MB)",
                start.elapsed().as_secs_f64(),
                sum / every as f64,
                bytes as f64 / 1e6,
            );
            sum = 0.0;
        }
    }
}
```

`VARY=0` — constant shape, flat forever:

```
step   200 t=  6s    28.21 ms/step  allocs    57 (8.4 MB)   <- includes JIT
step   400 t=  8s     9.83 ms/step  allocs    57 (8.4 MB)
step   600 t= 10s     9.62 ms/step  allocs    57 (8.4 MB)
```

`VARY=1 CYCLE=512` — 512 distinct shapes, then they repeat:

```
step   100 t=  5s    47.83 ms/step  allocs   555 (8.4 MB)   <- includes JIT
step   200 t=  6s    13.18 ms/step  allocs  1055 (8.5 MB)
step   300 t=  8s    16.55 ms/step  allocs  1555 (8.5 MB)
step   400 t= 10s    21.86 ms/step  allocs  2055 (8.5 MB)
step   500 t= 12s    23.49 ms/step  allocs  2555 (8.5 MB)
step   600 t= 15s    28.82 ms/step  allocs  2613 (8.5 MB)   <- shapes repeat, growth stops
step   700 t= 18s    26.33 ms/step  allocs  2613 (8.5 MB)
step   900 t= 23s    25.25 ms/step  allocs  2613 (8.5 MB)
```

`VARY=1 CYCLE=2048` — shapes keep changing, so it grows until the cache is full
(4096 entries + a ~51-allocation baseline), and stays there:

```
step   300 t=  8s    27.05 ms/step  allocs  1555 (8.5 MB)
step   600 t= 16s    24.94 ms/step  allocs  3055 (8.5 MB)
step   900 t= 27s    36.65 ms/step  allocs  4147 (8.5 MB)
step  1200 t= 39s    39.82 ms/step  allocs  4147 (8.5 MB)
step  1800 t= 66s    46.62 ms/step  allocs  4147 (8.5 MB)
step  3000 t=115s    38.86 ms/step  allocs  4147 (8.5 MB)
```

A 4× slowdown, 9.6 → 39–47 ms/step, on identical work. (Past the cap the step time stops
climbing and drifts inside that band — LRU eviction keeps the slice count fixed but not the
layout. Every run here has the CPU governor pinned to `powersave`; the absolute ms move with
it, the growth does not.)

Live allocations rise by exactly 5 per step — one slice per launch whose info words are new
— while `bytes_in_use` barely moves, because that figure is the sum of the slices' *requested*
sizes and an info buffer is only tens of bytes; the 512 B each one occupies in the page is
padding, counted in `bytes_padding`/`bytes_reserved`. The step time tracks the slice count,
not the bytes.

*What the live slices actually are.* Instrumenting `MemoryManagement::reserve` to dump a
size histogram of every live slice, per pool (`VARY=1 CYCLE=2048`, one line per 5000
reservations; pool 1 is the 24.24 MB-page sliced pool above):

```
reserve #10000  pool 1:  375 live slices: 512B x324, 1024B x25, 262144B x24, 8192B x1, 2097152B x1
reserve #40000  pool 1: 1358 live slices: 512B x1289, 1024B x24, 262144B x24, 2097152B x17, ...
reserve #70000  pool 1: 2313 live slices: 512B x2250, 1024B x31, 262144B x30, 8192B x1, 2097152B x1
reserve #100000 pool 1: 3263 live slices: 512B x3210, 1024B x26, 262144B x25, 8192B x1, 2097152B x1
```

Only the 512 B class grows; every other size is a steady working set. The 2 MB activations
sit in the same pool, so they pay for the 3210 cached info buffers in front of them.

**Turning the cache off removes it entirely.** A one-line patch adding an env override to
`MetadataCachePolicy::default`, then the same binary and workload (`VARY=1 CYCLE=2048`):

| step | live allocs (default / cache off) | ms/step (default) | ms/step (cache off) |
|---|---|---|---|
| 300 | 1555 / 51 | 26.0 | 22.4 (both include JIT) |
| 600 | 3055 / 51 | 24.1 | 9.6 |
| 900 | 4147 / 51 | 35.8 | 9.6 |
| 1200 | 4147 / 51 | 36.7 | 9.8 |

With the cache off, the varying-shape run is exactly as fast as the constant-shape one
(`VARY=0` with the cache off: 9.7 ms/step) and stays there.

**A workaround, for anyone who hits this.** burn's `Backend::memory_cleanup(&device)` —
or cubecl's `ComputeClient::memory_cleanup`, which is what it calls and what I measured —
reaches
[`MetadataInfoCache::clear_unpinned`](https://github.com/tracel-ai/cubecl/blob/2a3c997dafc43081d58c1417111c35c726dc46d6/crates/cubecl-server/src/metadata_cache.rs#L249),
which drops every entry no live graph pins. Calling it every 100 steps in the loop above
holds the repro at 51 live allocations and **10.5 ms/step, flat** — against 39 ms and
climbing without it, and 9.6 ms with the cache compiled out, so the call itself costs
~0.8 ms per step here. It is a blunt instrument (it also returns free pages to the driver),
but it works today.

**Where the time goes.** Sampling profile (gperftools, 500 Hz) of a real training run
(a Mamba-style LM, cubecl `fc902af`), 100 steps early (step 100–200) vs 100 steps later
(step 600–700) — 9097 vs 14647 samples for the same number of steps:

| function | early | late |
|---|---|---|
| `Client::launch` (self — inside the channel's `enqueue`, i.e. waiting on the device thread) | 1405 | 3795 |
| `MemoryPage::coalesce` (incl. callees) | 1046 | 2834 |
| `MemoryPage::try_reserve` (incl. callees) | 808 | 2213 |
| `ManagedMemoryDescriptor::update_slice` | 532 | 1551 |
| `Client::do_empty` (self — same `enqueue`) | 584 | 1555 |
| `ManagedMemoryHandle::is_free` | 312 | 789 |

The device thread's stacks are `MemoryManagement::reserve` → `SlicedPool::try_reserve` →
`MemoryPage::coalesce`, reached from `Server::initialize_memory` (ordinary tensor
allocations). Per-phase host time in that run went from `fwd 35 / bwd 84 / opt 103` ms at
step 100 to `fwd 63 / bwd 133 / opt 256` ms at step 600, and GPU utilization fell from 33%
to 17% because the GPU waits on the host.

*The same model, with and without the cache* (AdamW, identical config, live allocations
from `memory_usage`, `pool 1` from the same per-pool dump):

| run | step 100 | step 325 | step 425 |
|---|---|---|---|
| cache on | 167 ms, 915 allocs (653 in pool 1) | 229 ms, 1594 (1332) | — (run ended) |
| cache off | 147 ms, 405 allocs (143 in pool 1) | 148 ms, 405 (143) | 152 ms, 405 (143) |

Bytes in use stayed at ~41 MB in both. It reproduces on cubecl `fc902af`
(burn `8f08fdad`) and `2a3c997` (burn `3a93fbfc`) alike.

**Expected behavior**

Step time should not depend on how many *distinct* shapes a process has seen. A cache
that holds 4096 entries nothing will read again is bad on its own, but the sharp edge is
that holding them makes every unrelated allocation slower.

Possible directions, in the order they helped here:

1. **Don't let the cache pin pool slices.** Serving info buffers from their own arena
   (or a pool whose reservation cost does not depend on occupancy) would keep a full
   cache from taxing every other allocation.
2. **Make the reservation path not O(live slices).** `coalesce` and `add_new_slice`
   renumbering every slice on every reservation is the dominant term; coalescing only when
   something was freed since the last pass, or keeping free slices in a size-indexed
   structure, would bound it.
3. **Make the cache adaptive.** A workload that keeps missing gains nothing from 4096
   retained entries; a hit-rate check, or not caching info whose shape words are seen for
   the first time, would keep the steady-state small. (`clear_unpinned` already exists and
   is enough to recover — see the workaround above — it is just never reached on its own.)

**Environment**

- cubecl `2a3c997` (via burn `3a93fbfc`) and cubecl `fc902af` (via burn `8f08fdad`)
- RTX 2060 (6 GB), CUDA 12.x
- rustc 1.97.1, release build; CPU governor pinned to `powersave` for every timing run
